crates/red_knot_python_semantic/src/semantic_index/definition.rs

```diff
@@ -1,57 +1,133 @@
+                DefinitionKind::Alias(AstNodeRef::new(parsed, alias))
+            }
+            DefinitionNodeRef::Function(function) => {
+                DefinitionKind::Function(AstNodeRef::new(parsed, function))
+            }
+            DefinitionNodeRef::Class(class) => {
+                DefinitionKind::Class(AstNodeRef::new(parsed, class))
+            }
+            DefinitionNodeRef::NamedExpression(named) => {
+                DefinitionKind::NamedExpression(AstNodeRef::new(parsed, named))
+            }
+            DefinitionNodeRef::Target(target) => {
+                DefinitionKind::Target(AstNodeRef::new(parsed, target))
+            }
+        }
     }
 }
 
-impl From<ScopedClassId> for Definition {
-    fn from(value: ScopedClassId) -> Self {
-        Self::ClassDef(value)
+impl DefinitionNodeRef<'_> {
+    pub(super) fn key(self) -> DefinitionNodeKey {
+        match self {
+            Self::Alias(node) => DefinitionNodeKey::from(node),
+            Self::Function(node) => DefinitionNodeKey::from(node),
+            Self::Class(node) => DefinitionNodeKey::from(node),
+            Self::NamedExpression(node) => DefinitionNodeKey::from(node),
+            Self::Target(node) => DefinitionNodeKey(NodeKey::from_node(node)),
+        }
     }
 }
 
-impl From<ScopedFunctionId> for Definition {
-    fn from(value: ScopedFunctionId) -> Self {
-        Self::FunctionDef(value)
+#[derive(Clone, Debug)]
+pub enum DefinitionKind {
+    Alias(AstNodeRef<ast::Alias>),
+    Function(AstNodeRef<ast::StmtFunctionDef>),
+    Class(AstNodeRef<ast::StmtClassDef>),
+    NamedExpression(AstNodeRef<ast::ExprNamed>),
+    Target(AstNodeRef<ast::Expr>),
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub(super) struct DefinitionNodeKey(NodeKey);
+
+impl From<&ast::Alias> for DefinitionNodeKey {
+    fn from(node: &ast::Alias) -> Self {
+        Self(NodeKey::from_node(node))
     }
 }
 
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
-pub struct ImportDefinition {
-    pub(crate) import_id: ScopedStatementId,
+impl From<&ast::StmtFunctionDef> for DefinitionNodeKey {
+    fn from(node: &ast::StmtFunctionDef) -> Self {
+        Self(NodeKey::from_node(node))
+    }
+}
 
-    /// Index into [`ruff_python_ast::StmtImport::names`].
-    pub(crate) alias: u32,
+impl From<&ast::StmtClassDef> for DefinitionNodeKey {
+    fn from(value: &ast::StmtClassDef) -> Self {
+        Self(NodeKey::from_node(value))
+    }
 }
 
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
-pub struct ImportFromDefinition {
-    pub(crate) import_id: ScopedStatementId,
+impl From<&ast::ExprNamed> for DefinitionNodeKey {
+    fn from(value: &ast::ExprNamed) -> Self {
+        Self(NodeKey::from_node(value))
+    }
+}
 
-    /// Index into [`ruff_python_ast::StmtImportFrom::names`].
-    pub(crate) name: u32,
+impl From<&ast::ExprName> for DefinitionNodeKey {
+    fn from(value: &ast::ExprName) -> Self {
+        Self(NodeKey::from_node(value))
+    }
```

Comment:
Is this `From` impl used? Why implement `From` just for `ExprName` and not other exprs?

Reply:
You're an observant reviewer! No, it isn't.